Objects/dictobject.c

```diff
@@ -2510,6 +2562,28 @@ PyDict_Copy(PyObject *o)
             _PyObject_GC_TRACK(split_copy);
         return (PyObject *)split_copy;
     }
+
+    if (PyDict_CheckExact(mp) &&
+            mp->ma_values == NULL && mp->ma_keys->dk_refcnt == 1 &&
```

Comment:
dk_refcnt==1 test seems redundant, the inner assertion is enough IMHO.

Extract of dictobject.c:

```
The DictObject can be in one of two forms.

Either:
  A combined table:
    ma_values == NULL, dk_refcnt == 1.
    Values are stored in the me_value field of the PyDictKeysObject.
Or:
  A split table:
    ma_values != NULL, dk_refcnt >= 1
    Values are stored in the ma_values array.
    Only string (unicode) keys are allowed.
    All dicts sharing same key must have same insertion order.
```

dk_refcnt can only be != 1 if ma_values is not NULL.